File: DSCI553/hw3/task1.py
import sys
from pyspark import SparkContext
import time
import random
from itertools import combinations

if __name__ == '__main__':
    input_file_path = sys.argv[1]
    output_file_path = sys.argv[2]
    
    start_time = time.time()
    
    sc = SparkContext(appName='task1_lsh')
    sc.setLogLevel('ERROR')
    
    data_lines = sc.textFile(input_file_path)
    header = data_lines.first()
    data_lines = data_lines.filter(lambda line: line != header).map(lambda line: line.split(','))
    
    business_users_rdd = data_lines.map(lambda row: (row[1], row[0])).groupByKey().mapValues(set)
    
    business_users_dict = {}
    for biz, users in business_users_rdd.collect():
        business_users_dict[biz] = users
    
    all_users = data_lines.map(lambda row: row[0]).distinct()
    user_to_index = {}
    index = 0
    for user in all_users.collect():
        user_to_index[user] = index
        index += 1
    
    num_hashes = 60
    num_users = index
    large_prime = int(1e9 + 7)
    
    random.seed(42)
    coeff_a = random.sample(range(1, num_users), num_hashes)
    coeff_b = random.sample(range(1, num_users), num_hashes)
    
    signatures = {}
    for biz, user_set in business_users_rdd.collect():
        signature_list = []
        for hash_idx in range(num_hashes):
            min_hash_value = float('inf')
            for user in user_set:
                hash_result = ((coeff_a[hash_idx] * user_to_index[user] + coeff_b[hash_idx]) % large_prime) % num_users
                min_hash_value = min(min_hash_value, hash_result)
            signature_list.append(int(min_hash_value))
        signatures[biz] = signature_list
    
    rows_per_band = 2
    num_bands = num_hashes // rows_per_band
    
    buckets = {}
    for biz, sig in signatures.items():
        for band_idx in range(num_bands):
            band_signature = tuple(sig[band_idx * rows_per_band : (band_idx + 1) * rows_per_band])
            bucket_id = (band_idx, band_signature)
            if bucket_id not in buckets:
                buckets[bucket_id] = []
            buckets[bucket_id].append(biz)
    
    filtered_buckets = {}
    for bucket_id, biz_list in buckets.items():
        if len(biz_list) > 1:
            filtered_buckets[bucket_id] = biz_list
    
    candidate_pairs = set()
    for biz_list in filtered_buckets.values():
        sorted_list = sorted(biz_list)
        for pair in combinations(sorted_list, 2):
            candidate_pairs.add(pair)
    
    results = {}
    for biz1, biz2 in candidate_pairs:
        users1 = business_users_dict[biz1]
        users2 = business_users_dict[biz2]
        intersection = len(users1 & users2)
        union = len(users1 | users2)
        jaccard_sim = intersection / union
        
        if jaccard_sim >= 0.5:
            key = f"{biz1},{biz2}"
            results[key] = jaccard_sim
    
    results = dict(sorted(results.items()))
    
    output_content = "business_id_1, business_id_2, similarity\n"
    for key, sim in results.items():
        output_content += f"{key},{sim}\n"
    
    with open(output_file_path, 'w') as f:
        f.write(output_content)
    
    sc.stop()
    
    end_time = time.time()
    print('Duration:', end_time - start_time)

File: DSCI553/hw3/task1.scala
import org.apache.spark.{SparkConf, SparkContext}
import scala.util.Random
import java.io.PrintWriter

object Task1 {
  def main(args: Array[String]): Unit = {
    val inputPath  = args(0)  
    val outputPath = args(1)

    val conf = new SparkConf().setAppName("task1_lsh_scala")
    val sc   = new SparkContext(conf)
    sc.setLogLevel("ERROR")

    val lines = sc.textFile(inputPath)
    val header = lines.first()
    val ub = lines.filter(_ != header).map(_.split(",")).map(x => (x(0), x(1)))

    val bizUsers = ub.map{ case(u,b) => (b,u) }
      .groupByKey()
      .mapValues(_.toSet)
      .cache()

    val userToIndex = ub.map(_._1).distinct().zipWithIndex().mapValues(_.toInt).collectAsMap()
    val userToIndexBc = sc.broadcast(userToIndex)
    val numUsers = userToIndex.size

    val numHashes = 60
    val rowsPerBand = 2
    val numBands = numHashes / rowsPerBand
    val largePrime = 1000000007L
    val rnd = new Random(42)

    def sampleCoeffs(n: Int, upTo: Int): Array[Int] = {
      val limit = math.max(upTo, n + 1)
      rnd.shuffle((1 until limit).toList).take(n).toArray
    }
    val a = sampleCoeffs(numHashes, math.max(2, numUsers))
    val b = sampleCoeffs(numHashes, math.max(2, numUsers))

    val signatures = bizUsers.map { case (biz, users) =>
      val idxs = users.iterator.map(u => userToIndexBc.value.getOrElse(u, -1)).filter(_ >= 0).toArray
      val sig = Array.fill[Int](numHashes)(Int.MaxValue)
      var h = 0
      while (h < numHashes) {
        var minv = Int.MaxValue
        val ah = a(h).toLong
        val bh = b(h).toLong
        var i = 0
        while (i < idxs.length) {
          val x = idxs(i).toLong
          val hv = (((ah * x + bh) % largePrime) % math.max(1, numUsers)).toInt
          if (hv < minv) minv = hv
          i += 1
        }
        sig(h) = if (minv == Int.MaxValue) 0 else minv
        h += 1
      }
      (biz, sig)
    }.cache()

    val bandPairs = signatures.flatMap { case (biz, sig) =>
      (0 until numBands).iterator.map { band =>
        val start = band * rowsPerBand
        val key = (band, (sig(start), sig(start + 1))) 
        (key, biz)
      }
    }
      .groupByKey()
      .values
      .flatMap { bizIter =>
        val arr = bizIter.toArray.distinct.sorted
        for {
          i <- arr.indices.iterator
          j <- (i + 1) until arr.length
        } yield (arr(i), arr(j))
      }
      .distinct()

    val bizUsersMap = bizUsers.collectAsMap()
    val bizUsersBc  = sc.broadcast(bizUsersMap)

    val candidates = bandPairs.map { case (b1, b2) =>
      val s1 = bizUsersBc.value.getOrElse(b1, Set.empty[String])
      val s2 = bizUsersBc.value.getOrElse(b2, Set.empty[String])
      val inter = (s1 intersect s2).size.toDouble
      val uni   = (s1 union s2).size.toDouble
      val jac   = if (uni == 0) 0.0 else inter / uni
      (b1, b2, jac)
    }
      .filter(_._3 >= 0.5)
      .collect()
      .sortBy{ case (b1, b2, _) => (b1, b2) }

    val out = new PrintWriter(outputPath)
    out.println("business_id_1,business_id_2,similarity") 
    candidates.foreach { case (b1, b2, s) =>
      out.println(s"$b1,$b2,${"%.6f".format(s)}")
    }
    out.close()

    sc.stop()
  }
}